DSpace/dspace-angular
# DS-4283: (Google Scholar) Provide Entity Type Label/Name as part of URL path

## References
* Fixes #745
* Depends on #975
* Affected by REST issue https://github.com/DSpace/DSpace/issues/3113

## Description
An additional routing option has been added to support retrieving entities through their entity type and ID.
Visiting an item page will automatically redirect the user to this new entity route (when the item contains a value for `relationship.type`).
The new entity route is now also used instead of the item route wherever possible (e.g. item links, buttons...), the same goes for child-routes like `/edit`.
Furthermore, this new routing option will be used when redirecting handles.

## Instructions for Reviewers
This PR seems a lot bigger than it is, because it depends on #975 You can see only the changes in this PR [here](https://github.com/atmire/dspace-angular/compare/cache-redesign-part-2...atmire:Item-page-redirects)


### To test the route

* Navigate to an entity by going to an url with the following structure: `/entities/:entitytype/:uuid`
* Verify that this displays the corresponding entity.
Note: The `:entitytype` parameter is currently not used for the actual resolved of the entity, only the `:uuid`.

### To test the handle redirection

To test this, some mocks will need to be enabled in the code since the REST handle endpoint current does not work for retrieving items based on the handle, see https://github.com/DSpace/DSpace/issues/3113
To test an entity:

* Enable the `mockPublicationResponse` by uncommenting it in `src/app/shared/mocks/dspace-rest/mocks/response-map.mock.ts`
* Navigate to a handle url like `/handle/123456789/5`, the exact handle does not matter since the response will be overwritten by the mock response.
* Verify that you get redirected to `/entities/publication/0136fe00-43ca-439e-ae6c-b2b08c2b8f5e`.
To test a non-entity item: 
* Enable the `mockUntypedItemResponse` by uncommenting it in `src/app/shared/mocks/dspace-rest/mocks/response-map.mock.ts`
* Navigate to a handle url like `/handle/123456789/5`, the exact handle does not matter since the response will be overwritten by the mock response.
* Verify that you get redirected to `/items/0136fe00-43ca-439e-ae6c-b2b08c2b8f5e`.

### To test the item route redirect

* Visit the `/items/:uuid` route for an item containing a metadata value for `relationship.type`. Verify you're redirected to `/entities/:entitytype/:uuid` (also verify the entity type is lowercased)
* Visit the `/items/:uuid` route for an item containing **no** metadata value for `relationship.type`. Verify you still end up on the item page without entity route.

### To test entity route links

Any location containing a clickable link to an item page, should now contain a link to the entity route where available.

* Visit discovery and filter on items containing a relationship type. Hover over their links and verify they contain the entity route instead of the item route.
* Log in as administrator and visit the item page for an item containing a metadata value for `relationship.type`. Hover over the "edit item" button (top-right of the page) and  verify they contain the entity route instead of the item route. Clicking the button should redirect you to the edit page of the item on the entities route.
* Try to find other locations with item links and verify they translate to entity routes for items containing a metadata value for `relationship.type`.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
